plugins/linux-capture/pipewire.c

```diff
@@ -1138,12 +1167,20 @@ void obs_pipewire_destroy(obs_pipewire_data *obs_pw)
 	teardown_pipewire(obs_pw);
 	destroy_session(obs_pw);
 
+	g_clear_pointer(&obs_pw->restore_token, bfree);
+
 	bfree(obs_pw);
 }
 
+void obs_pipewire_save(obs_pipewire_data *obs_pw, obs_data_t *settings)
+{
+	obs_data_set_string(settings, "RestoreToken", obs_pw->restore_token);
```

Comment:
Could we use an array/object instead of a single string indexed by the first entry of `XDG_CURRENT_DESKTOP` (can be a :-seperated list). This way we could save a restore token for each DE making it possible to switch them without loosing the state each time. This can be usefull for simple setups with scenes for different outputs without creating a new set of scenes for each DE.

Reply:
Clients doesn't have access to that information, and that was a fundamental part of the protocol design. This exchange is strictly between xdg-desktop-portal and the portal implementation.

Reply:
does flatpak hide a `getenv("XDG_CURRENT_DESKTOP")`?

Reply:
Sorry, I'm not adding one restore token per desktop environment. It doesn't matter if we have access to environment variables or not. What you're proposing is merely a bad hack, OBS is the wrong layer to add this.

Reply:
Well on wlroots based applications using this envvar is actually a valid solution to eg. load settings wrt. the running compositor. But if there is no easy way to represent dynamic key-value pairs with obs_data_t I'm fine with solving this issue with profiles or other means.

Reply:
The point is not how easy or hard it would be to implement this in OBS Studio. What I called "bad hack" is doing this goes directly against the purpose of portals. Portals exist to have DE-independent API that apps can use regardless of where they are being executed. And yet here we are, talking about adding DE-specific code in a DE-independent code. So I'm 100% against adding DE-specific code here because that's the wrong place to add that - the correct place is in portal frontends. That's actually what you suggested when you reviewed the restore-token PR. Did you change your mind about It?

Reply:
>  the correct place is in portal frontends. That's actually what you suggested when you reviewed the restore-token PR. Did you change your mind about It?

If I understood it correctly a token is a 1:1 correspondence with a given entry in the permission-store. This token is only valid for one use. After it was used the portal implementation might issue a new token for an identical entry in the permission store. To support mutliple implementations we either need:
* making the token 1:N in the permission store and updating all entries to the new issued token.
* Let the client manage multiple tokens and use the corresponding one.

From the discussion at the restore-token MR I took the assumption that the 1:1 correspondence was important to keep.

> And yet here we are, talking about adding DE-specific code in a DE-independent code.
I'm not talking about DE-specific code. I'm talking about a DE-specific config entry using the same independent codepath.

Reply:
> If I understood it correctly a token is a 1:1 correspondence with a given entry in the permission-store. This token is only valid for one use. After it was used the portal implementation might issue a new token for an identical entry in the permission store.

None of this matters to applications. Applications express intent ("I want to restore session this later"), and get a blob to use next time. That's it. Nothing else matters for applications. It doesn't matter what the blob is. It doesn't matter how portals implement it. It doesn't matter that there are different frontends. This is all an implementation detail of portals, and the API design guarantees that none of this ever leaks into application code.

> I'm talking about a DE-specific config entry using the same independent codepath.

That's exactly the same. If you're writing a portals-based app and you have to check which DE the app is running on, then we failed at the API design level.

My question still stands: why did you request a way for portal frontends to exchange data between each other (the `(suv)` variant)? What you're asking here is the exact opposite of that, and I fail to see what your goals could even be when you request the exact opposite in each PR.

Reply:
> why did you request a way for portal frontends to exchange data between each other (the (suv) variant)? 

I requested the (suv) variant to have a more robust input validation in the portal implementation when receiving the blob from the permission-store. With only (v) collisions between the variants of a portal implementation could happen. With (suv) s has to be the name of the portal implementation. (Input validation has to be in place to prevent malice actors going against the spec).  

Parsing the variant of another portal was sth. undesired iirc so interoperability on the implementation side is out.

> What you're asking here is the exact opposite of that, and I fail to see what your goals could even be when you request the exact opposite in each PR.

My goal is to be able to configure DE1 in a specific way (including DE specific settings on different applications), switch to DE2, configure it in a completely different way, switch back to DE1 and be able to continue from where i left DE1. 

I don't care if that is sth. implemented 
* in the application (storing multiple tokens) or 
* in the portal frontend (saving multiple permission-store entries with the same token). 

We could also tell people to 
* use different obs profiles for different DEs 

and every screencast instance will happily save it's token. But as soon as obs starts with the wrong profile all tokens will probably be invalidated and the state of the scene/profile wrt. screencast will be lost.

I could solve this for me probably by using a wrapper
```
#!/bin/sh
obs --profile="$(echo $XDG_CURRENT_DESKTOP | cut -d: -f1)"
```
but loosing your setup because obs started with the "wrong" profile is not nice UX.

I don't know the obs userbase good enough if that is really an issue. I just wanted to make this issue visible.

Reply:
Note: This should not block this MR in any way. 
This is a huge improvement for users using only a single DE, which is the majority of users!
I'm asking if we want to consider this an issue with a solution now or later.

Reply:
> I requested the (suv) variant to have a more robust input validation in the portal implementation when receiving the blob from the permission-store. With only (v) collisions between the variants of a portal implementation could happen. With (suv) s has to be the name of the portal implementation. (Input validation has to be in place to prevent malice actors going against the spec).

There is literally no way any application could use the private data of this protocol to create any attack vector.

> My goal is to be able to configure DE1 in a specific way (including DE specific settings on different applications), switch to DE2, configure it in a completely different way, switch back to DE1 and be able to continue from where i left DE1.

It seems to me that this is a premature optimization to cater for an unrealistic use case. People don't change DEs often, and when they do, OBS is going to be the smallest of their problems. With this PR, selecting a monitor becomes a one-time action.

Reply:
> There is literally no way any application could use the private data of this protocol to create any attack vector.

This is only relevant between portal implementations, not applications, the same way I wanted the (suv) var for easier parsing in the implementation.

> People don't change DEs often

Fine for me. (Discussion resolved, can't see the button to collapse this.)

Reply:
> We could also tell people to
>    use different obs profiles for different DEs

This is not possible, profiles hold non-source/scene configuration. In this case the settings concerned are for the source which are contained in the scene collection.

Reply:
> This is not possible, profiles hold non-source/scene configuration. In this case the settings concerned are for the source which are contained in the scene collection.

Thanks. 
So on the script `s/profile/collection/`

Reply:
Let's do it like this: if we get a reasonable number of complaints (either through Discord, or the bug tracker here) that the monitor is queried when changing DEs, I'll implement @columbarius' suggestion. Otherwise, we can deem this as not an issue. What do you think?